### PR TITLE
expose profile events counters in system.part_log

### PR DESCRIPTION
**Pull request by bharatnc**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Expose ProfileEvents counters in system.part_log.

Aims to close: #35989  which supersedes #34314.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
We need this :)
Let's continue?

**Comment by bharatnc** [comment]
> We need this :) Let's continue?

yeah sure, it's in my planned work - I'll try to re-prioritize.

File: src/Storages/MergeTree/IExecutableTask.h [hunk 1]
```diff
@@ -30,6 +30,8 @@ class IExecutableTask
 public:
     using TaskResultCallback = std::function<void(bool)>;
     virtual bool executeStep() = 0;
+    virtual void onSuspend() {}
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/IExecutableTask.h:33]: These methods look unneeded, we can put them inside `executeStep`.

File: src/Storages/MergeTree/TaskObserverMetrics.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <Common/ThreadStatus.h>
+
+namespace DB
+{
+
+class TaskObserverMetrics : public boost::noncopyable
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/TaskObserverMetrics.h:8]: Missing comments.

Comment (alexey-milovidov) [inline src/Storages/MergeTree/TaskObserverMetrics.h:8]: Wrong name. It's not only about metrics.

**Comment by alexey-milovidov** [comment]
Test needed.

**Comment by alexey-milovidov** [comment]
Nice, it works:

```
SELECT
    P.Names,
    P.Values
FROM system.part_log
ARRAY JOIN ProfileEvents AS P

Query id: 5caa25f1-a201-4165-928c-ce31f549a881

┌─P.Names─────────────────────────────────┬────P.Values─┐
│ Query                                   │           1 │
│ FileOpen                                │         429 │
│ Seek                                    │         205 │
│ ReadBufferFromFileDescriptorRead        │       14097 │
│ ReadBufferFromFileDescriptorReadBytes   │ 14622071296 │
│ WriteBufferFromFileDescriptorWrite      │       54187 │
│ WriteBufferFromFileDescriptorWriteBytes │ 10692421783 │
│ ReadCompressedBytes                     │ 14441851816 │
│ CompressedReadBufferBlocks              │      595715 │
│ CompressedReadBufferBytes               │ 66443604393 │
│ OpenedFileCacheMisses                   │         211 │
│ IOBufferAllocs                          │         961 │
│ IOBufferAllocBytes                      │   455418791 │
│ FunctionExecute                         │       71515 │
│ CreatedReadBufferOrdinary               │         106 │
│ CreatedReadBufferDirectIO               │         105 │
│ DiskReadElapsedMicroseconds             │     5064495 │
│ DiskWriteElapsedMicroseconds            │     6001593 │
│ Merge                                   │       71516 │
│ MergedRows                              │   100000000 │
│ MergedUncompressedBytes                 │ 74687843978 │
│ MergesTimeMilliseconds                  │       11087 │
│ ContextLock                             │         316 │
│ RWLockAcquiredReadLocks                 │           1 │
│ RealTimeMicroseconds                    │   178704084 │
│ UserTimeMicroseconds                    │   167843822 │
│ SystemTimeMicroseconds                  │     6385909 │
│ SoftPageFaults                          │       85062 │
│ OSCPUWaitMicroseconds                   │       15880 │
│ OSCPUVirtualTimeMicroseconds            │   174229731 │
│ OSReadBytes                             │ 14581051392 │
│ OSWriteBytes                            │ 10693091328 │
│ OSReadChars                             │ 14622071749 │
│ OSWriteChars                            │ 10692665911 │
│ QueryProfilerRuns                       │         354 │
└─────────────────────────────────────────┴─────────────┘

35 rows in set. Elapsed: 0.002 sec.
```

**Comment by alexey-milovidov** [comment]
It still does not work for background operations.

**Comment by bharatnc** [comment]
@alexey-milovidov thank you for fixes and sorry I haven't been able to get this PR working yet.

If I understand this code from my ex co-worker, it still needs work as it won't probably work with background tasks like you have mentioned.

We want this feature too, but I don't think that I will be able to get to it immediately (not in the next couple of weeks).

**Comment by alexey-milovidov** [comment]
No problem and sorry for the preliminary changes :)
We will merge it soon.

**Comment by elenatorro** [comment]
Hi,
These logs would be very useful. Would it be possible to know if there's a plan on including them in the short term? Thanks!

File: src/Interpreters/ProfileEventsExt.cpp [hunk 3]
```diff
@@ -19,6 +19,22 @@ std::shared_ptr<DB::DataTypeEnum8> TypeEnum = std::make_shared<DB::DataTypeEnum8
     { "gauge",     static_cast<Int8>(GAUGE)},
 });
 
+String dumpToString(const Counters::Snapshot & counters, bool nonzero_only)
```
Comment (vdimir) [inline src/Interpreters/ProfileEventsExt.cpp:22]: Used for debug, need to remove

File: src/Interpreters/PartLog.cpp [hunk 4]
```diff
@@ -233,4 +253,28 @@ bool PartLog::addNewParts(
...
+
+    return addNewParts(context, part_log_entries, execution_status);
+}
+
+bool PartLog::addNewPart(ContextPtr context, const PartLog::MutableDataPartPtr & part, UInt64 elapsed_ns, const ExecutionStatus & execution_status)
+{
+    PartLog::PartLogEntries part_log_entries;
+    part_log_entries.emplace_back(part, elapsed_ns);
+    return addNewParts(context, part_log_entries, execution_status);
+}
```
Comment (vdimir) [inline src/Interpreters/PartLog.cpp:278]: These overloads should be removed. I'll check where else they're used and replace them with new ones.

**Comment by alesapin** [comment]
Sorry, forget about it... will fix it.

File: src/Common/ThreadStatus.h [hunk 5]
```diff
@@ -107,6 +107,7 @@ using ThreadGroupStatusPtr = std::shared_ptr<ThreadGroupStatus>;
  * - https://github.com/ClickHouse/ClickHouse/pull/40078
  */
 extern thread_local constinit ThreadStatus * current_thread;
+extern thread_local constinit ProfileEvents::Counters * subthread_profile_events;
```
Comment (vdimir) [inline src/Common/ThreadStatus.h:110]: Probably it's worth moving this pointer can be moved into `ThreadStatus current_thread`  not to introduce one more thread_local variable.

Comment (alesapin) [reply src/Common/ThreadStatus.h:110]: Agree, would be simpler.

**Review by alesapin — APPROVED** [review]
Looks Ok.

**Comment by alexey-milovidov** [comment]
@vdimir, there are no resource usage counters:

```
milovidov-desktop :) SELECT PE.Names, PE.Values FROM system.part_log ARRAY JOIN ProfileEvents AS PE WHERE table = 'uk_price_paid' AND event_type = 'MergeParts' ORDER BY event_time DESC

SELECT
    PE.Names,
    PE.Values
FROM system.part_log
ARRAY JOIN ProfileEvents AS PE
WHERE (table = 'uk_price_paid') AND (event_type = 'MergeParts')
ORDER BY event_time DESC

Query id: 6b4ab24f-e070-4e5d-83be-97737337e1de

┌─PE.Names────────────────────────────────┬──PE.Values─┐
│ FileOpen                                │        194 │
│ ReadBufferFromFileDescriptorRead        │        320 │
│ ReadBufferFromFileDescriptorReadBytes   │  216292400 │
│ WriteBufferFromFileDescriptorWrite      │        608 │
│ WriteBufferFromFileDescriptorWriteBytes │  179080584 │
│ ReadCompressedBytes                     │  216292400 │
│ CompressedReadBufferBlocks              │      11893 │
│ CompressedReadBufferBytes               │  834374886 │
│ OpenedFileCacheMisses                   │        147 │
│ IOBufferAllocs                          │        470 │
│ IOBufferAllocBytes                      │  119823558 │
│ CreatedReadBufferOrdinary               │        147 │
│ DiskReadElapsedMicroseconds             │     137308 │
│ DiskWriteElapsedMicroseconds            │     127250 │
│ Merge                                   │       3415 │
│ MergedRows                              │   27910954 │
│ MergedUncompressedBytes                 │ 2692905763 │
│ MergesTimeMilliseconds                  │      14704 │
│ ContextLock                             │        314 │
│ QueryProfilerRuns                       │         33 │
└─────────────────────────────────────────┴────────────┘
```

Like UserTimeMicroseconds, OSCPUVirtualTimeMicroseconds, etc.

**Pull request merged**